api/cluster/resource/templater.go

```diff
@@ -550,6 +596,120 @@ func createLoggerSpec(loggerURL string, loggerConfig models.LoggerConfig) *kserv
 	}
 }
 
+// createNewInferenceServiceTopologySpreadConstraints creates topology spread constrains for a component of a new
+// inference service
+func createNewInferenceServiceTopologySpreadConstraints(
+	modelService *models.Service,
+	config *config.DeploymentConfig,
+	component kservev1beta1.ComponentType,
+) ([]corev1.TopologySpreadConstraint, error) {
+	if len(config.TopologySpreadConstraints) == 0 {
+		var topologySpreadConstraints []corev1.TopologySpreadConstraint
+		return topologySpreadConstraints, nil
+	}
+	var newRevisionName string
+	if modelService.DeploymentMode == deployment.RawDeploymentMode {
+		newRevisionName = fmt.Sprintf("isvc.%s-%s-default", modelService.Name, component)
+	} else if modelService.DeploymentMode == deployment.ServerlessDeploymentMode ||
+		modelService.DeploymentMode == deployment.EmptyDeploymentMode {
+		newRevisionName = fmt.Sprintf("%s-%s-default-00001", modelService.Name, component)
+	} else {
+		return nil, fmt.Errorf("invalid deployment mode: %s", modelService.DeploymentMode)
+	}
+	return appendPodSpreadingLabelSelectorsToTopologySpreadConstraints(
+		config.TopologySpreadConstraints,
+		newRevisionName,
+	)
+}
+
+// updateExistingInferenceServiceTopologySpreadConstraints creates topology spread constraints for a component of a new
+// inference service
+func updateExistingInferenceServiceTopologySpreadConstraints(
+	orig *kservev1beta1.InferenceService,
+	modelService *models.Service,
+	config *config.DeploymentConfig,
+	component kservev1beta1.ComponentType,
+) ([]corev1.TopologySpreadConstraint, error) {
+	if len(config.TopologySpreadConstraints) == 0 {
+		var topologySpreadConstraints []corev1.TopologySpreadConstraint
+		return topologySpreadConstraints, nil
+	}
+	var newRevisionName string
+	if modelService.DeploymentMode == deployment.RawDeploymentMode {
+		newRevisionName = fmt.Sprintf("isvc.%s-%s-default", modelService.Name, component)
+	} else if modelService.DeploymentMode == deployment.ServerlessDeploymentMode ||
+		modelService.DeploymentMode == deployment.EmptyDeploymentMode {
+		var err error
+		newRevisionName, err = getNewRevisionNameForExistingSeverlessDeployment(
+			orig.Status.Components[component].LatestReadyRevision,
```

Comment:
Hmm.. What if there was a failed deployment or one in progress? Should we get the latest "created" revision as opposed to ready ?

```
      Latest Created Revision:      std-preprocess-loadtest-3-predictor-default-00002
      Latest Ready Revision:        std-preprocess-loadtest-3-predictor-default-00002
      Latest Rolledout Revision:    std-preprocess-loadtest-3-predictor-default-00002
      Previous Rolledout Revision:  std-preprocess-loadtest-3-predictor-default-00001
```

Reply:
@deadlycoconuts  / @ariefrahmansyah any thoughts on this?

Reply:
Ah oops my bad; I totally forgot about this. Hmm so if I understand correctly, the latest created revision >= latest rolled out revision and we trying to be as 'safe' as possible by picking the largest possible revision number to increment right? I think that sounds good. We can monitor the behaviour and if it ever creates any bugs/unexpected behaviour, change the implementation later on :)